backend/danswer/server/danswer_api.py

```diff
@@ -0,0 +1,205 @@
+    if token != saved_key or not saved_key:
+        raise HTTPException(status_code=401, detail="Invalid API key")
+    return token
+
+
+# Provides a way to recover if the api key is deleted for some reason
+# Can also just restart the server to regenerate a new one
+def api_key_dep_if_exist(authorization: str | None = Header(None)) -> str | None:
+    token = authorization.removeprefix("Bearer ").strip() if authorization else None
+    saved_key = get_danswer_api_key(dont_regenerate=True)
+    if not saved_key:
+        return None
+
+    if token != saved_key:
+        raise HTTPException(status_code=401, detail="Invalid API key")
+
+    return token
+
+
+@router.post("/regenerate-key")
+def regenerate_key(_: str | None = Depends(api_key_dep_if_exist)) -> ApiKey:
+    delete_danswer_api_key()
+    return ApiKey(api_key=cast(str, get_danswer_api_key()))
+
+
+@router.post("/doc-ingestion")
+def document_ingestion(
+    doc_info: IngestionDocument,
+    _: str = Depends(api_key_dep),
+    db_session: Session = Depends(get_session),
+) -> IngestionResult:
+    """Currently only attaches docs to existing connectors (cc-pairs).
+    Or to the default ingestion connector that is accessible to all users
+
+    Things to note:
+    - The document id if not provided is automatically generated from the semantic identifier
+      so if the document source type etc is updated, it won't create a duplicate
+    """
+    if doc_info.credential_id:
+        credential_id = doc_info.credential_id
+        credential = fetch_credential_by_id(
+            credential_id=credential_id,
+            user=None,
+            db_session=db_session,
+            assume_admin=True,
+        )
+        if credential is None:
+            raise ValueError("Invalid Credential for doc, does not exist.")
+    else:
+        credential_id = 0
+
+    connector_id = doc_info.connector_id
+    # If user provides id and name, id takes precedence
+    if connector_id is not None:
+        connector = fetch_connector_by_id(connector_id, db_session)
+        if connector is None:
+            raise ValueError("Invalid Connector for doc, id does not exist.")
+    elif doc_info.connector_name:
+        connector = fetch_ingestion_connector_by_name(
+            doc_info.connector_name, db_session
+        )
+        if connector is None:
+            raise ValueError("Invalid Connector for doc, name does not exist.")
+        connector_id = connector.id
+    else:
+        connector_id = 0
+
+    cc_pair = get_connector_credential_pair(
+        connector_id=connector_id, credential_id=credential_id, db_session=db_session
+    )
+    if cc_pair is None:
+        raise ValueError("Connector and Credential not associated.")
+
+    # Disregard whatever value is passed, this must be True
+    doc_info.document.from_ingestion_api = True
+
+    document = Document.from_base(doc_info.document)
+
+    # TODO once the frontend is updated with this enum, remove this logic
```

Comment:
Actually, this PR just can't be put in without the frontend supporting the new enum, I could maybe just do it but, might as well just wait until the frontend is fully in